File: app/src/domain/gating.ts
/**
 * Deterministic fault gating (spec §5.9). The cloud worker is the source of truth
 * for the FaultEvaluation it writes onto a SwingAnalysis, but the same logic lives
 * here so the app can reason about gates consistently and so it is unit-testable.
 *
 * A gate either opens or it does not; severity is a band, not an LLM judgement. The
 * AI may only ever select a fault whose gate is open.
 */

import { type MetricMeta } from './faultLibrary';
import type {
  FaultEvaluation,
  FaultGate,
  Metric,
  MetricStatus,
} from './types';

/** Confidence below this withholds a fault / softens the score. */
export const LOW_CONFIDENCE = 0.5;

/** Clamp helper. */
function clamp(v: number, lo: number, hi: number): number {
  return Math.max(lo, Math.min(hi, v));
}

/** Is a measured value physically plausible for its metric? */
export function assessMetricStatus(value: number, meta: MetricMeta): MetricStatus {
  if (Number.isNaN(value)) return 'implausible';
  if (value < meta.plausible.min || value > meta.plausible.max) return 'implausible';
  return 'ok';
}

export function isInFriendlyRange(value: number, meta: MetricMeta): boolean {
  return value >= meta.friendlyRange.min && value <= meta.friendlyRange.max;
}

/**
 * Evaluate a gate's condition against a value. Returns whether it fired and a
 * normalised magnitude (0 at the threshold, growing past it), used for severity and
 * for prioritising between multiple open gates.
 */
export function evaluateGate(
  value: number,
  gate: FaultGate,
): { fired: boolean; magnitude: number } {
  const { operator, threshold } = gate;
  switch (operator) {
    case 'exceeds': {
      const t = threshold.min ?? 0;
      const fired = value > t;
      // normalise by the threshold's own scale so different metrics compare.
      const magnitude = fired ? (value - t) / Math.max(Math.abs(t), 1) : 0;
      return { fired, magnitude };
    }
    case 'below': {
      const t = threshold.max ?? 0;
      const fired = value < t;
      const magnitude = fired ? (t - value) / Math.max(Math.abs(t), 1) : 0;
      return { fired, magnitude };
    }
    case 'outside': {
      const lo = threshold.min ?? -Infinity;
      const hi = threshold.max ?? Infinity;
      const fired = value < lo || value > hi;
      const dist = value < lo ? lo - value : value > hi ? value - hi : 0;
      const scale = Math.max(Math.abs(lo), Math.abs(hi), 1);
      return { fired, magnitude: fired ? dist / scale : 0 };
    }
    case 'magnitude_at_least': {
      const t = threshold.value ?? 0;
      const fired = Math.abs(value) >= t;
      const magnitude = fired ? (Math.abs(value) - t) / Math.max(t, 1) : 0;
      return { fired, magnitude };
    }
    default: {
      const exhaustive: never = operator;
      return exhaustive;
    }
  }
}

export function severityBand(magnitude: number): FaultEvaluation['severityBand'] {
  if (magnitude <= 0) return 'none';
  if (magnitude < 0.4) return 'mild';
  if (magnitude < 1.0) return 'moderate';
  return 'strong';
}

/** Find a metric on an analysis by key. */
export function getMetric(metrics: Metric[], key: string): Metric | undefined {
  return metrics.find((m) => m.key === key);
}

/**
 * Among the evaluated faults, pick the priority fault: the highest
 * magnitude × severityWeight among gates that fired with adequate confidence AND are
 * claim-eligible (the validation gate — spec §13.1). A fired-but-`soft_only` fault is
 * recorded for transparency but never becomes the primary claim. Returns null when
 * nothing clears the bar (→ the AI has nothing to select; the report shows "no clear
 * priority fault" rather than highlighting a crude proxy).
 */
export function pickPrimaryFault(
  evaluations: FaultEvaluation[],
  severityWeightOf: (faultId: string) => number,
): FaultEvaluation | null {
  const eligible = evaluations.filter(
    (e) => e.fired && e.confidence >= LOW_CONFIDENCE && e.status === 'ok' && e.claimEligible,
  );
  if (eligible.length === 0) return null;
  return eligible.reduce((best, e) => {
    const score = e.magnitude * severityWeightOf(e.faultId);
    const bestScore = best.magnitude * severityWeightOf(best.faultId);
    return score > bestScore ? e : best;
  });
}

/**
 * Compose a metric's confidence from worker engine confidence × the metric's
 * reliability tier, used by both gating and the score's confidence gate.
 */
export function metricConfidence(engineConfidence: number, meta: MetricMeta): number {
  const reliabilityFactor =
    meta.reliabilityTag === 'reliable' ? 1.0 : meta.reliabilityTag === 'approximate' ? 0.7 : 0.0;
  return clamp(engineConfidence * reliabilityFactor, 0, 1);
}

File: worker/src/swingsight_worker/domain/gating.py
"""Deterministic fault gating — the SAME logic as app/src/domain/gating.ts.

A gate either opens or it does not; severity is a band, not a judgement. The AI may
only ever select a fault whose gate is open. The app reasons about gates with the TS
version of this file; the worker is the source of truth that writes the
FaultEvaluation. They must agree, so every formula here mirrors gating.ts exactly —
do not "improve" one side without the other.
"""

from __future__ import annotations

import math
from typing import Callable, Literal, Optional

from .fault_library import FaultGate, MetricMeta

# Confidence below this withholds a fault / softens the score (mirror LOW_CONFIDENCE).
LOW_CONFIDENCE = 0.5

MetricStatus = Literal["ok", "unmeasurable_view", "low_confidence", "implausible"]
SeverityBand = Literal["none", "mild", "moderate", "strong"]


def clamp(v: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, v))


def assess_metric_status(value: float, meta: MetricMeta) -> MetricStatus:
    """Is a measured value physically plausible for its metric? (mirror)."""
    if math.isnan(value):
        return "implausible"
    if value < meta.plausible.min or value > meta.plausible.max:
        return "implausible"
    return "ok"


def is_in_friendly_range(value: float, meta: MetricMeta) -> bool:
    return meta.friendly_range.min <= value <= meta.friendly_range.max


def evaluate_gate(value: float, gate: FaultGate) -> tuple[bool, float]:
    """Evaluate a gate against a value -> (fired, magnitude). magnitude is 0 at the
    threshold and grows past it, normalised by the threshold's own scale so different
    metrics compare. Mirror of evaluateGate (operator-for-operator)."""
    operator = gate.operator
    threshold = gate.threshold
    if operator == "exceeds":
        t = threshold.get("min", 0) or 0
        fired = value > t
        magnitude = (value - t) / max(abs(t), 1) if fired else 0.0
        return fired, magnitude
    if operator == "below":
        t = threshold.get("max", 0) or 0
        fired = value < t
        magnitude = (t - value) / max(abs(t), 1) if fired else 0.0
        return fired, magnitude
    if operator == "outside":
        lo = threshold.get("min", -math.inf)
        hi = threshold.get("max", math.inf)
        fired = value < lo or value > hi
        dist = (lo - value) if value < lo else (value - hi) if value > hi else 0.0
        scale = max(abs(lo) if lo != -math.inf else 0, abs(hi) if hi != math.inf else 0, 1)
        return fired, (dist / scale if fired else 0.0)
    if operator == "magnitude_at_least":
        t = threshold.get("value", 0) or 0
        fired = abs(value) >= t
        magnitude = (abs(value) - t) / max(t, 1) if fired else 0.0
        return fired, magnitude
    raise ValueError(f"unknown gate operator: {operator}")


def severity_band(magnitude: float) -> SeverityBand:
    """Mirror of severityBand: none/mild/moderate/strong by magnitude."""
    if magnitude <= 0:
        return "none"
    if magnitude < 0.4:
        return "mild"
    if magnitude < 1.0:
        return "moderate"
    return "strong"


def metric_confidence(engine_confidence: float, meta: MetricMeta) -> float:
    """Compose a metric's confidence from engine confidence × its reliability tier.
    Mirror of metricConfidence."""
    factor = (
        1.0
        if meta.reliability_tag == "reliable"
        else 0.7
        if meta.reliability_tag == "approximate"
        else 0.0
    )
    return clamp(engine_confidence * factor, 0.0, 1.0)


def pick_primary_fault(
    evaluations: list[dict],
    severity_weight_of: Callable[[str], float],
) -> Optional[dict]:
    """Among the evaluated faults, pick the priority fault: highest
    magnitude × severityWeight among gates that fired with adequate confidence, an 'ok'
    status, AND claim-eligible (the validation gate — spec §13.1; a fired-but-soft_only
    fault is recorded but never the primary claim). Returns None when nothing clears the
    bar. Mirror of pickPrimaryFault (the worker writes the result; the app reasons
    identically)."""
    eligible = [
        e
        for e in evaluations
        if e["fired"]
        and e["confidence"] >= LOW_CONFIDENCE
        and e["status"] == "ok"
        and e.get("claimEligible", True)
    ]
    if not eligible:
        return None
    best = eligible[0]
    best_score = best["magnitude"] * severity_weight_of(best["faultId"])
    for e in eligible[1:]:
        score = e["magnitude"] * severity_weight_of(e["faultId"])
        if score > best_score:
            best, best_score = e, score
    return best


def pick_observation_fault(
    evaluations: list[dict],
    severity_weight_of: Callable[[str], float],
) -> Optional[dict]:
    """Among the evaluated faults, pick the top TENTATIVE observation: same
    magnitude × severityWeight ranking as pick_primary_fault, but the INVERSE claim
    filter — a fired, adequately-confident, plausible fault that is NOT claim-eligible
    (a soft_only proxy). This is what the coaching layer may hedge on when no
    claim-eligible primary cleared the bar; it is never the primary claim. Returns None
    when nothing qualifies. Mirror of pickObservationFault."""
    eligible = [
        e
        for e in evaluations
        if e["fired"]
        and e["confidence"] >= LOW_CONFIDENCE
        and e["status"] == "ok"
        and not e.get("claimEligible", True)
    ]
    if not eligible:
        return None
    best = eligible[0]
    best_score = best["magnitude"] * severity_weight_of(best["faultId"])
    for e in eligible[1:]:
        score = e["magnitude"] * severity_weight_of(e["faultId"])
        if score > best_score:
            best, best_score = e, score
    return best
